// mcp-echarts/build/tools/bar.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateBarChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Bar chart data schema
const data = zod_1.z.object({
    category: zod_1.z
        .string()
        .describe("Category of the data point, such as 'Category A'."),
    value: zod_1.z.number().describe("Value of the data point, such as 10."),
    group: zod_1.z
        .string()
        .optional()
        .describe("Group name for multiple series, used for grouping or stacking"),
});
exports.generateBarChartTool = {
    name: "generate_bar_chart",
    description: "Generate a bar chart to show data for numerical comparisons among different categories, such as, comparing categorical data and for horizontal comparisons.",
    inputSchema: zod_1.z.object({
        axisXTitle: schema_1.AxisXTitleSchema,
        axisYTitle: schema_1.AxisYTitleSchema,
        data: zod_1.z
            .array(data)
            .describe("Data for bar chart, such as, [{ category: 'Category A', value: 10 }, { category: 'Category B', value: 20 }] or [{ category: 'Category A', value: 10, group: 'Group A' }].")
            .nonempty({ message: "Bar chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        group: zod_1.z
            .boolean()
            .optional()
            .default(false)
            .describe("Whether grouping is enabled. When enabled, bar charts require a 'group' field in the data. When `group` is true, `stack` should be false."),
        stack: zod_1.z
            .boolean()
            .optional()
            .default(false)
            .describe("Whether stacking is enabled. When enabled, bar charts require a 'group' field in the data. When `stack` is true, `group` should be false."),
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { axisXTitle, axisYTitle, data, height, group = false, stack = false, theme, title, width, outputType, } = params;
        // Check if data has group field for multiple series
        const hasGroups = data.some((item) => item.group);
        let series = [];
        let categories = [];
        if (hasGroups && (group || stack)) {
            // Handle multiple series data (grouped or stacked)
            const groupMap = new Map();
            const categorySet = new Set();
            // Group data by group field and collect all categories
            for (const item of data) {
                const groupName = item.group || "Default";
                if (!groupMap.has(groupName)) {
                    groupMap.set(groupName, []);
                }
                const groupData = groupMap.get(groupName);
                if (groupData) {
                    groupData.push({ category: item.category, value: item.value });
                }
                categorySet.add(item.category);
            }
            // Sort categories
            categories = Array.from(categorySet).sort();
            // Create series for each group
            groupMap.forEach((groupData, groupName) => {
                // Create a map for quick lookup
                const dataMap = new Map(groupData.map((d) => [d.category, d.value]));
                // Fill values for all categories (0 for missing data)
                const values = categories.map((category) => { var _a; return (_a = dataMap.get(category)) !== null && _a !== void 0 ? _a : 0; });
                series.push({
                    data: values,
                    name: groupName,
                    stack: stack ? "Total" : undefined,
                    type: "bar",
                });
            });
        }
        else {
            // Handle single series data
            categories = data.map((item) => item.category);
            const values = data.map((item) => item.value);
            series = [
                {
                    data: values,
                    type: "bar",
                },
            ];
        }
        const echartsOption = {
            legend: hasGroups && (group || stack)
                ? {
                    left: "center",
                    orient: "horizontal",
                    bottom: 10,
                }
                : undefined,
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "axis",
            },
            xAxis: {
                data: categories,
                name: axisXTitle,
                type: "category",
            },
            yAxis: {
                name: axisYTitle,
                type: "value",
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_bar_chart");
    }),
};
