// mcp-echarts/build/tools/boxplot.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateBoxplotChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Boxplot chart data schema
const data = zod_1.z.object({
    category: zod_1.z
        .string()
        .describe("Category of the data point, such as 'Category A'."),
    value: zod_1.z.number().describe("Value of the data point, such as 10."),
    group: zod_1.z
        .string()
        .optional()
        .describe("Optional group for the data point, used for grouping in the boxplot."),
});
exports.generateBoxplotChartTool = {
    name: "generate_boxplot_chart",
    description: "Generate a boxplot chart to show data for statistical summaries among different categories, such as, comparing the distribution of data points across categories.",
    inputSchema: zod_1.z.object({
        axisXTitle: schema_1.AxisXTitleSchema,
        axisYTitle: schema_1.AxisYTitleSchema,
        data: zod_1.z
            .array(data)
            .describe("Data for boxplot chart, such as, [{ category: 'Category A', value: 10 }, { category: 'Category B', value: 20, group: 'Group A' }].")
            .nonempty({ message: "Boxplot chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        var _a, _b, _c;
        const { axisXTitle, axisYTitle, data, height, theme, title, width, outputType, } = params;
        // Group data by category and optional group
        const hasGroups = data.some((item) => item.group);
        let categories = [];
        const boxplotData = [];
        if (hasGroups) {
            // Group by category and group
            const groupMap = new Map();
            const categorySet = new Set();
            for (const item of data) {
                const groupName = item.group || "Default";
                const key = `${item.category}_${groupName}`;
                if (!groupMap.has(key)) {
                    groupMap.set(key, []);
                }
                (_a = groupMap.get(key)) === null || _a === void 0 ? void 0 : _a.push(item);
                categorySet.add(item.category);
            }
            categories = Array.from(categorySet).sort();
            // Calculate boxplot statistics for each group
            groupMap.forEach((values, key) => {
                const [category, group] = key.split("_");
                const sortedValues = values.map((v) => v.value).sort((a, b) => a - b);
                const boxplotStats = calculateBoxplotStats(sortedValues);
                boxplotData.push({
                    name: `${category}-${group}`,
                    value: boxplotStats,
                });
            });
        }
        else {
            // Group by category only
            const categoryMap = new Map();
            for (const item of data) {
                if (!categoryMap.has(item.category)) {
                    categoryMap.set(item.category, []);
                }
                (_b = categoryMap.get(item.category)) === null || _b === void 0 ? void 0 : _b.push(item.value);
            }
            categories = Array.from(categoryMap.keys()).sort();
            // Calculate boxplot statistics for each category
            for (const category of categories) {
                const values = ((_c = categoryMap.get(category)) === null || _c === void 0 ? void 0 : _c.sort((a, b) => a - b)) || [];
                const boxplotStats = calculateBoxplotStats(values);
                boxplotData.push({
                    name: category,
                    value: boxplotStats,
                });
            }
        }
        const series = [
            {
                type: "boxplot",
                data: boxplotData,
                itemStyle: {
                    borderWidth: 2,
                },
                emphasis: {
                    itemStyle: {
                        borderWidth: 3,
                        shadowBlur: 5,
                        shadowColor: "rgba(0, 0, 0, 0.3)",
                    },
                },
            },
        ];
        const echartsOption = {
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "item",
            },
            xAxis: {
                type: "category",
                data: categories,
                name: axisXTitle,
                boundaryGap: true,
                nameGap: 30,
                splitArea: {
                    show: false,
                },
                splitLine: {
                    show: false,
                },
            },
            yAxis: {
                type: "value",
                name: axisYTitle,
                splitArea: {
                    show: true,
                },
            },
        };
        // Helper function to calculate boxplot statistics
        function calculateBoxplotStats(values) {
            const len = values.length;
            const min = values[0];
            const max = values[len - 1];
            const median = len % 2 === 0
                ? (values[len / 2 - 1] + values[len / 2]) / 2
                : values[Math.floor(len / 2)];
            const q1Index = Math.floor(len / 4);
            const q3Index = Math.floor((3 * len) / 4);
            const Q1 = values[q1Index];
            const Q3 = values[q3Index];
            return [min, Q1, median, Q3, max];
        }
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_boxplot_chart");
    }),
};
