// mcp-echarts/build/tools/gauge.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateGaugeChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Gauge chart data schema
const data = zod_1.z.object({
    name: zod_1.z.string().describe("Indicator name, such as 'CPU Usage'."),
    value: zod_1.z.number().describe("Current value of the indicator, such as 75."),
});
exports.generateGaugeChartTool = {
    name: "generate_gauge_chart",
    description: "Generate a gauge chart to display single indicator's current status, such as, CPU usage rate, completion progress, or performance scores.",
    inputSchema: zod_1.z.object({
        data: zod_1.z
            .array(data)
            .describe("Data for gauge chart, such as, [{ name: 'CPU Usage', value: 75 }]. Multiple gauges can be displayed.")
            .nonempty({ message: "Gauge chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        max: zod_1.z
            .number()
            .optional()
            .default(100)
            .describe("Maximum value of the gauge, default is 100."),
        min: zod_1.z
            .number()
            .optional()
            .default(0)
            .describe("Minimum value of the gauge, default is 0."),
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, max = 100, min = 0, theme, title, width, outputType, } = params;
        // For multiple gauges, arrange them horizontally
        const series = data.map((item, index) => {
            const isMultiple = data.length > 1;
            return {
                name: item.name,
                type: "gauge",
                data: [{ name: item.name, value: item.value }],
                center: isMultiple
                    ? [`${(100 / (data.length + 1)) * (index + 1)}%`, "60%"]
                    : ["50%", "55%"],
                radius: isMultiple ? `${Math.min(80 / data.length, 30)}%` : "80%",
                min: min,
                max: max,
                startAngle: 180,
                endAngle: 0,
                axisLine: {
                    lineStyle: {
                        width: 6,
                        color: [
                            [0.3, "#67e0e3"],
                            [0.7, "#37a2da"],
                            [1, "#fd666d"],
                        ],
                    },
                },
                pointer: {
                    itemStyle: {
                        color: "inherit",
                    },
                },
                axisTick: {
                    distance: -30,
                    length: 8,
                    lineStyle: {
                        color: "#fff",
                        width: 2,
                    },
                },
                splitLine: {
                    distance: -30,
                    length: 30,
                    lineStyle: {
                        color: "#fff",
                        width: 4,
                    },
                },
                axisLabel: {
                    color: "inherit",
                    distance: 40,
                    fontSize: isMultiple ? 10 : 12,
                },
                detail: {
                    valueAnimation: true,
                    formatter: "{value}",
                    color: "inherit",
                    fontSize: isMultiple ? 16 : 20,
                    offsetCenter: [0, "30%"],
                },
                title: {
                    offsetCenter: [0, "50%"],
                    fontSize: isMultiple ? 12 : 14,
                },
            };
        });
        const echartsOption = {
            legend: data.length > 1
                ? {
                    bottom: 10,
                    left: "center",
                    orient: "horizontal",
                    data: data.map((item) => item.name),
                }
                : undefined,
            series,
            title: {
                left: "center",
                text: title,
                top: data.length > 1 ? "5%" : undefined,
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_gauge_chart");
    }),
};
