// mcp-echarts/build/tools/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTreeChartTool = exports.generateParallelChartTool = exports.generateGraphChartTool = exports.generateBoxplotChartTool = exports.generateCandlestickChartTool = exports.generateHeatmapChartTool = exports.generateSunburstChartTool = exports.generateTreemapChartTool = exports.generateGaugeChartTool = exports.generateFunnelChartTool = exports.generateSankeyChartTool = exports.generateScatterChartTool = exports.generateRadarChartTool = exports.generatePieChartTool = exports.generateBarChartTool = exports.generateLineChartTool = exports.generateEChartsTool = exports.tools = void 0;
const bar_1 = require("./bar.js");
Object.defineProperty(exports, "generateBarChartTool", { enumerable: true, get: function () { return bar_1.generateBarChartTool; } });
const boxplot_1 = require("./boxplot.js");
Object.defineProperty(exports, "generateBoxplotChartTool", { enumerable: true, get: function () { return boxplot_1.generateBoxplotChartTool; } });
const candlestick_1 = require("./candlestick.js");
Object.defineProperty(exports, "generateCandlestickChartTool", { enumerable: true, get: function () { return candlestick_1.generateCandlestickChartTool; } });
const echarts_1 = require("./echarts.js");
Object.defineProperty(exports, "generateEChartsTool", { enumerable: true, get: function () { return echarts_1.generateEChartsTool; } });
const funnel_1 = require("./funnel.js");
Object.defineProperty(exports, "generateFunnelChartTool", { enumerable: true, get: function () { return funnel_1.generateFunnelChartTool; } });
const gauge_1 = require("./gauge.js");
Object.defineProperty(exports, "generateGaugeChartTool", { enumerable: true, get: function () { return gauge_1.generateGaugeChartTool; } });
const graph_1 = require("./graph.js");
Object.defineProperty(exports, "generateGraphChartTool", { enumerable: true, get: function () { return graph_1.generateGraphChartTool; } });
const heatmap_1 = require("./heatmap.js");
Object.defineProperty(exports, "generateHeatmapChartTool", { enumerable: true, get: function () { return heatmap_1.generateHeatmapChartTool; } });
const line_1 = require("./line.js");
Object.defineProperty(exports, "generateLineChartTool", { enumerable: true, get: function () { return line_1.generateLineChartTool; } });
const parallel_1 = require("./parallel.js");
Object.defineProperty(exports, "generateParallelChartTool", { enumerable: true, get: function () { return parallel_1.generateParallelChartTool; } });
const pie_1 = require("./pie.js");
Object.defineProperty(exports, "generatePieChartTool", { enumerable: true, get: function () { return pie_1.generatePieChartTool; } });
const radar_1 = require("./radar.js");
Object.defineProperty(exports, "generateRadarChartTool", { enumerable: true, get: function () { return radar_1.generateRadarChartTool; } });
const sankey_1 = require("./sankey.js");
Object.defineProperty(exports, "generateSankeyChartTool", { enumerable: true, get: function () { return sankey_1.generateSankeyChartTool; } });
const scatter_1 = require("./scatter.js");
Object.defineProperty(exports, "generateScatterChartTool", { enumerable: true, get: function () { return scatter_1.generateScatterChartTool; } });
const sunburst_1 = require("./sunburst.js");
Object.defineProperty(exports, "generateSunburstChartTool", { enumerable: true, get: function () { return sunburst_1.generateSunburstChartTool; } });
const tree_1 = require("./tree.js");
Object.defineProperty(exports, "generateTreeChartTool", { enumerable: true, get: function () { return tree_1.generateTreeChartTool; } });
const treemap_1 = require("./treemap.js");
Object.defineProperty(exports, "generateTreemapChartTool", { enumerable: true, get: function () { return treemap_1.generateTreemapChartTool; } });
exports.tools = [
    echarts_1.generateEChartsTool,
    line_1.generateLineChartTool,
    bar_1.generateBarChartTool,
    pie_1.generatePieChartTool,
    radar_1.generateRadarChartTool,
    scatter_1.generateScatterChartTool,
    sankey_1.generateSankeyChartTool,
    funnel_1.generateFunnelChartTool,
    gauge_1.generateGaugeChartTool,
    treemap_1.generateTreemapChartTool,
    sunburst_1.generateSunburstChartTool,
    heatmap_1.generateHeatmapChartTool,
    candlestick_1.generateCandlestickChartTool,
    boxplot_1.generateBoxplotChartTool,
    graph_1.generateGraphChartTool,
    parallel_1.generateParallelChartTool,
    tree_1.generateTreeChartTool,
];
