// mcp-echarts/build/tools/radar.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateRadarChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Radar chart data schema
const data = zod_1.z.object({
    name: zod_1.z.string().describe("Dimension name, such as 'Design'."),
    value: zod_1.z.number().describe("Value of the dimension, such as 70."),
    group: zod_1.z
        .string()
        .optional()
        .describe("Group name for multiple series, used for comparing different entities"),
});
exports.generateRadarChartTool = {
    name: "generate_radar_chart",
    description: "Generate a radar chart to display multidimensional data (four dimensions or more), such as, evaluate Huawei and Apple phones in terms of five dimensions: ease of use, functionality, camera, benchmark scores, and battery life.",
    inputSchema: zod_1.z.object({
        data: zod_1.z
            .array(data)
            .describe("Data for radar chart, such as, [{ name: 'Design', value: 70 }, { name: 'Performance', value: 85 }] or [{ name: 'Design', value: 70, group: 'iPhone' }].")
            .nonempty({ message: "Radar chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, theme, title, width, outputType } = params;
        // Check if data has group field for multiple series
        const hasGroups = data.some((item) => item.group);
        // Collect all unique dimensions
        const dimensionSet = new Set();
        for (const item of data) {
            dimensionSet.add(item.name);
        }
        const dimensions = Array.from(dimensionSet).sort();
        // Create radar indicator configuration
        // Calculate the maximum value for all dimensions, then use the same max value to avoid alignTicks warning
        const allValues = data.map((item) => item.value);
        const globalMaxValue = Math.max(...allValues);
        const unifiedMax = Math.ceil((globalMaxValue * 1.2) / 10) * 10;
        const indicator = dimensions.map((name) => ({
            name,
            max: unifiedMax,
        }));
        let series = [];
        if (hasGroups) {
            // Handle multiple series data (grouped)
            const groupMap = new Map();
            // Group data by group field
            for (const item of data) {
                const groupName = item.group || "Default";
                if (!groupMap.has(groupName)) {
                    groupMap.set(groupName, []);
                }
                const groupData = groupMap.get(groupName);
                if (groupData) {
                    groupData.push({ name: item.name, value: item.value });
                }
            }
            // Create series data for each group
            const seriesData = Array.from(groupMap.entries()).map(([groupName, groupData]) => {
                // Create a map for quick lookup
                const dataMap = new Map(groupData.map((d) => [d.name, d.value]));
                // Fill values for all dimensions (0 for missing data)
                const values = dimensions.map((dimension) => { var _a; return (_a = dataMap.get(dimension)) !== null && _a !== void 0 ? _a : 0; });
                return {
                    name: groupName,
                    value: values,
                };
            });
            series = [
                {
                    data: seriesData,
                    type: "radar",
                },
            ];
        }
        else {
            // Handle single series data
            const dataMap = new Map(data.map((d) => [d.name, d.value]));
            const values = dimensions.map((dimension) => { var _a; return (_a = dataMap.get(dimension)) !== null && _a !== void 0 ? _a : 0; });
            series = [
                {
                    data: [
                        {
                            value: values,
                            name: title || "Data",
                        },
                    ],
                    type: "radar",
                },
            ];
        }
        const echartsOption = {
            legend: hasGroups
                ? {
                    left: "center",
                    orient: "horizontal",
                    bottom: "5%",
                }
                : undefined,
            radar: {
                indicator,
                radius: "60%",
                splitNumber: 4,
                axisName: {
                    formatter: "{value}",
                    color: "#666",
                },
                splitArea: {
                    areaStyle: {
                        color: ["rgba(250, 250, 250, 0.3)", "rgba(200, 200, 200, 0.3)"],
                    },
                },
            },
            series,
            title: {
                left: "center",
                text: title,
                top: "5%",
            },
            tooltip: {
                trigger: "item",
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_radar_chart");
    }),
};
