// mcp-echarts/build/tools/sankey.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSankeyChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Sankey chart data schema
const data = zod_1.z.object({
    source: zod_1.z.string().describe("Source node name, such as 'Landing Page'."),
    target: zod_1.z.string().describe("Target node name, such as 'Product Page'."),
    value: zod_1.z
        .number()
        .describe("Flow value between source and target, such as 50000."),
});
exports.generateSankeyChartTool = {
    name: "generate_sankey_chart",
    description: "Generate a sankey chart to visualize the flow of data between different stages or categories, such as, the user journey from landing on a page to completing a purchase.",
    inputSchema: zod_1.z.object({
        data: zod_1.z
            .array(data)
            .describe("Data for sankey chart, such as, [{ source: 'Landing Page', target: 'Product Page', value: 50000 }, { source: 'Product Page', target: 'Add to Cart', value: 35000 }].")
            .nonempty({ message: "Sankey chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        nodeAlign: zod_1.z
            .enum(["left", "right", "justify"])
            .optional()
            .default("justify")
            .describe("Alignment of nodes in the sankey chart, such as, 'left', 'right', or 'justify'."),
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, nodeAlign = "justify", theme, title, width, outputType, } = params;
        // Extract unique nodes from data
        const nodeSet = new Set();
        for (const item of data) {
            nodeSet.add(item.source);
            nodeSet.add(item.target);
        }
        // Create nodes array
        const nodes = Array.from(nodeSet).map((name) => ({ name }));
        // Create links array
        const links = data.map((item) => ({
            source: item.source,
            target: item.target,
            value: item.value,
        }));
        const series = [
            {
                type: "sankey",
                data: nodes,
                links: links,
                emphasis: {
                    focus: "adjacency",
                },
                nodeAlign: nodeAlign,
                left: "10%",
                top: "10%",
                right: "10%",
                bottom: "10%",
                label: {
                    position: "right",
                    color: "#000",
                },
                lineStyle: {
                    color: "gradient",
                    curveness: 0.5,
                },
            },
        ];
        const echartsOption = {
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "item",
                triggerOn: "mousemove",
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_sankey_chart");
    }),
};
