// mcp-echarts/build/tools/scatter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateScatterChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Scatter chart data schema
const data = zod_1.z.object({
    x: zod_1.z.number().describe("X coordinate value of the data point."),
    y: zod_1.z.number().describe("Y coordinate value of the data point."),
});
exports.generateScatterChartTool = {
    name: "generate_scatter_chart",
    description: "Generate a scatter chart to show the relationship between two variables, helps discover their relationship or trends, such as, the strength of correlation, data distribution patterns.",
    inputSchema: zod_1.z.object({
        axisXTitle: schema_1.AxisXTitleSchema,
        axisYTitle: schema_1.AxisYTitleSchema,
        data: zod_1.z
            .array(data)
            .describe("Data for scatter chart, such as, [{ x: 10, y: 15 }, { x: 20, y: 25 }].")
            .nonempty({ message: "Scatter chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { axisXTitle, axisYTitle, data, height, theme, title, width, outputType, } = params;
        // Transform data for ECharts scatter chart
        const scatterData = data.map((item) => [item.x, item.y]);
        const series = [
            {
                data: scatterData,
                type: "scatter",
                symbolSize: 8,
                emphasis: {
                    focus: "series",
                    itemStyle: {
                        shadowBlur: 10,
                        shadowOffsetX: 0,
                        shadowColor: "rgba(0, 0, 0, 0.5)",
                    },
                },
            },
        ];
        const echartsOption = {
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "item",
            },
            xAxis: {
                name: axisXTitle,
                type: "value",
                scale: true,
            },
            yAxis: {
                name: axisYTitle,
                type: "value",
                scale: true,
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_scatter_chart");
    }),
};
