// mcp-echarts/build/tools/treemap.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTreemapChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Define recursive schema for hierarchical data
const TreeNodeSchema = zod_1.z.lazy(() => zod_1.z.object({
    name: zod_1.z.string().describe("Node name, such as 'Design'."),
    value: zod_1.z.number().describe("Node value, such as 70."),
    children: zod_1.z
        .array(TreeNodeSchema)
        .optional()
        .describe("Child nodes for hierarchical structure."),
}));
exports.generateTreemapChartTool = {
    name: "generate_treemap_chart",
    description: "Generate a treemap chart to display hierarchical data and can intuitively show comparisons between items at the same level, such as, show disk space usage with treemap.",
    inputSchema: zod_1.z.object({
        data: zod_1.z
            .array(TreeNodeSchema)
            .describe("Data for treemap chart, such as, [{ name: 'Design', value: 70, children: [{ name: 'Tech', value: 20 }] }].")
            .nonempty({ message: "Treemap chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, theme, title, width, outputType } = params;
        const series = [
            {
                type: "treemap",
                data: data,
                left: "3%",
                right: "3%",
                bottom: "3%",
                label: {
                    show: true,
                    formatter: "{b}",
                    fontSize: 12,
                    color: "#fff",
                },
                emphasis: {
                    focus: "descendant",
                    itemStyle: {
                        borderWidth: 3,
                    },
                    label: {
                        fontSize: 16,
                    },
                },
                breadcrumb: {
                    show: false,
                },
                roam: false,
                nodeClick: "zoomToNode",
            },
        ];
        const echartsOption = {
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "item",
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_treemap_chart");
    }),
};
