// mcp-echarts/build/tools/candlestick.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateCandlestickChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Candlestick chart data schema
const data = zod_1.z.object({
    date: zod_1.z.string().describe("Date string, such as '2023-01-01'."),
    open: zod_1.z.number().describe("Opening price."),
    high: zod_1.z.number().describe("Highest price."),
    low: zod_1.z.number().describe("Lowest price."),
    close: zod_1.z.number().describe("Closing price."),
    volume: zod_1.z.number().optional().describe("Trading volume (optional)."),
});
exports.generateCandlestickChartTool = {
    name: "generate_candlestick_chart",
    description: "Generate a candlestick chart for financial data visualization, such as, stock prices, cryptocurrency prices, or other OHLC (Open-High-Low-Close) data.",
    inputSchema: zod_1.z.object({
        data: zod_1.z
            .array(data)
            .describe("Data for candlestick chart, such as, [{ date: '2023-01-01', open: 100, high: 110, low: 95, close: 105, volume: 10000 }].")
            .nonempty({ message: "Candlestick chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        showVolume: zod_1.z
            .boolean()
            .optional()
            .default(false)
            .describe("Whether to show volume chart below candlestick. Default is false."),
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, showVolume = false, theme, title, width, outputType, } = params;
        // Sort data by date
        const sortedData = [...data].sort((a, b) => new Date(a.date).getTime() - new Date(b.date).getTime());
        // Extract dates and OHLC data
        const dates = sortedData.map((item) => item.date);
        const ohlcData = sortedData.map((item) => [
            item.open,
            item.close,
            item.low,
            item.high,
        ]);
        const volumeData = sortedData.map((item) => item.volume || 0);
        const series = [
            {
                name: "Candlestick",
                type: "candlestick",
                data: ohlcData,
                itemStyle: {
                    color: "#ef232a",
                    color0: "#14b143",
                    borderColor: "#ef232a",
                    borderColor0: "#14b143",
                },
                emphasis: {
                    itemStyle: {
                        color: "#ef232a",
                        color0: "#14b143",
                        borderColor: "#ef232a",
                        borderColor0: "#14b143",
                    },
                },
            },
        ];
        // Add volume series if requested
        if (showVolume && volumeData.some((v) => v > 0)) {
            series.push({
                name: "Volume",
                type: "bar",
                xAxisIndex: 1,
                yAxisIndex: 1,
                data: volumeData,
                barWidth: "60%",
                itemStyle: {
                    color: (params) => {
                        const dataIndex = params.dataIndex;
                        return sortedData[dataIndex].close >= sortedData[dataIndex].open
                            ? "#ef232a"
                            : "#14b143";
                    },
                },
            });
        }
        const echartsOption = {
            animation: false,
            legend: {
                bottom: 10,
                left: "center",
                data: showVolume ? ["Candlestick", "Volume"] : ["Candlestick"],
            },
            tooltip: {
                trigger: "axis",
                axisPointer: {
                    type: "cross",
                },
                borderWidth: 1,
                borderColor: "#ccc",
                padding: 10,
                textStyle: {
                    color: "#000",
                },
            },
            xAxis: [
                {
                    type: "category",
                    data: dates,
                    boundaryGap: true,
                    axisLine: { onZero: false },
                    splitLine: { show: false },
                    min: -0.2,
                    max: dates.length - 0.8,
                },
                ...(showVolume
                    ? [
                        {
                            type: "category",
                            gridIndex: 1,
                            data: dates,
                            boundaryGap: true,
                            axisLine: { onZero: false },
                            axisTick: { show: false },
                            splitLine: { show: false },
                            axisLabel: { show: false },
                            min: -0.2,
                            max: dates.length - 0.8,
                        },
                    ]
                    : []),
            ],
            yAxis: [
                {
                    scale: true,
                    splitArea: {
                        show: true,
                    },
                },
                ...(showVolume
                    ? [
                        {
                            scale: true,
                            gridIndex: 1,
                            splitNumber: 2,
                            axisLabel: { show: false },
                            axisLine: { show: false },
                            axisTick: { show: false },
                            splitLine: { show: false },
                            min: 0,
                        },
                    ]
                    : []),
            ],
            grid: showVolume
                ? [
                    {
                        left: "12%",
                        right: "10%",
                        top: "15%",
                        height: "50%",
                    },
                    {
                        left: "12%",
                        right: "10%",
                        top: "75%",
                        height: "15%",
                    },
                ]
                : [
                    {
                        left: "12%",
                        right: "10%",
                        top: "15%",
                        bottom: "15%",
                    },
                ],
            series,
            title: {
                left: "center",
                text: title,
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_candlestick_chart");
    }),
};
