// mcp-echarts/build/tools/funnel.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateFunnelChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Funnel chart data schema
const data = zod_1.z.object({
    category: zod_1.z
        .string()
        .describe("Stage category name, such as 'Browse Website'."),
    value: zod_1.z.number().describe("Value at this stage, such as 50000."),
});
exports.generateFunnelChartTool = {
    name: "generate_funnel_chart",
    description: "Generate a funnel chart to visualize the progressive reduction of data as it passes through stages, such as, the conversion rates of users from visiting a website to completing a purchase.",
    inputSchema: zod_1.z.object({
        data: zod_1.z
            .array(data)
            .describe("Data for funnel chart, such as, [{ category: 'Browse Website', value: 50000 }, { category: 'Add to Cart', value: 35000 }, { category: 'Generate Order', value: 25000 }].")
            .nonempty({ message: "Funnel chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, theme, title, width, outputType } = params;
        // Transform data for ECharts funnel chart
        const funnelData = data.map((item) => ({
            name: item.category,
            value: item.value,
        }));
        const series = [
            {
                type: "funnel",
                data: funnelData,
                left: "10%",
                top: 60,
                width: "80%",
                height: "80%",
                min: 0,
                max: Math.max(...data.map((item) => item.value)),
                minSize: "0%",
                maxSize: "100%",
                sort: "descending",
                gap: 2,
                label: {
                    show: true,
                    position: "inside",
                    fontSize: 12,
                    color: "#fff",
                },
                labelLine: {
                    length: 10,
                    lineStyle: {
                        width: 1,
                        type: "solid",
                    },
                },
                itemStyle: {
                    borderColor: "#fff",
                    borderWidth: 1,
                },
                emphasis: {
                    label: {
                        fontSize: 16,
                    },
                },
            },
        ];
        const echartsOption = {
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "item",
            },
            legend: {
                left: "center",
                orient: "horizontal",
                bottom: 10,
                data: funnelData.map((item) => item.name),
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_funnel_chart");
    }),
};
