// mcp-echarts/build/tools/heatmap.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateHeatmapChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Heatmap chart data schema
const data = zod_1.z.object({
    x: zod_1.z
        .union([zod_1.z.string(), zod_1.z.number()])
        .describe("X axis value, such as 'Mon' or 0."),
    y: zod_1.z
        .union([zod_1.z.string(), zod_1.z.number()])
        .describe("Y axis value, such as 'AM' or 0."),
    value: zod_1.z.number().describe("Heat value at this position, such as 5."),
});
exports.generateHeatmapChartTool = {
    name: "generate_heatmap_chart",
    description: "Generate a heatmap chart to display data density or intensity distribution, such as, user activity patterns by time and day, or correlation matrix.",
    inputSchema: zod_1.z.object({
        axisXTitle: schema_1.AxisXTitleSchema,
        axisYTitle: schema_1.AxisYTitleSchema,
        data: zod_1.z
            .array(data)
            .describe("Data for heatmap chart, such as, [{ x: 'Mon', y: '12AM', value: 5 }, { x: 'Tue', y: '1AM', value: 3 }].")
            .nonempty({ message: "Heatmap chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { axisXTitle, axisYTitle, data, height, theme, title, width, outputType, } = params;
        // Extract unique x and y values
        const xValues = Array.from(new Set(data.map((item) => item.x))).sort();
        const yValues = Array.from(new Set(data.map((item) => item.y))).sort();
        // Create data map for quick lookup
        const dataMap = new Map();
        for (const item of data) {
            dataMap.set(`${item.x}_${item.y}`, item.value);
        }
        // Transform data for ECharts heatmap
        const heatmapData = [];
        for (let i = 0; i < xValues.length; i++) {
            for (let j = 0; j < yValues.length; j++) {
                const value = dataMap.get(`${xValues[i]}_${yValues[j]}`) || 0;
                heatmapData.push([i, j, value]);
            }
        }
        // Calculate value range for visual map
        const values = data.map((item) => item.value);
        const minValue = Math.min(...values);
        const maxValue = Math.max(...values);
        const series = [
            {
                type: "heatmap",
                data: heatmapData,
                label: {
                    show: true,
                    fontSize: 10,
                },
                emphasis: {
                    itemStyle: {
                        shadowBlur: 10,
                        shadowColor: "rgba(0, 0, 0, 0.5)",
                    },
                },
            },
        ];
        const echartsOption = {
            grid: {
                height: "60%",
                top: "15%",
                right: "15%",
                bottom: "10%",
            },
            series,
            title: {
                left: "center",
                text: title,
                top: "3%",
            },
            tooltip: {
                position: "top",
            },
            visualMap: {
                min: minValue,
                max: maxValue,
                calculable: true,
                orient: "horizontal",
                left: "center",
                bottom: "15%",
                inRange: {
                    color: [
                        "#313695",
                        "#4575b4",
                        "#74add1",
                        "#abd9e9",
                        "#e0f3f8",
                        "#ffffcc",
                        "#fee090",
                        "#fdae61",
                        "#f46d43",
                        "#d73027",
                        "#a50026",
                    ],
                },
            },
            xAxis: {
                type: "category",
                data: xValues,
                name: axisXTitle,
                splitArea: {
                    show: true,
                },
            },
            yAxis: {
                type: "category",
                data: yValues,
                name: axisYTitle,
                splitArea: {
                    show: true,
                },
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_heatmap_chart");
    }),
};
