// mcp-echarts/build/tools/pie.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generatePieChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Pie chart data schema
const data = zod_1.z.object({
    category: zod_1.z
        .string()
        .describe("Category of the data point, such as 'Category A'."),
    value: zod_1.z.number().describe("Value of the data point, such as 27."),
});
exports.generatePieChartTool = {
    name: "generate_pie_chart",
    description: "Generate a pie chart to show the proportion of parts, such as, market share and budget allocation.",
    inputSchema: zod_1.z.object({
        data: zod_1.z
            .array(data)
            .describe("Data for pie chart, such as, [{ category: 'Category A', value: 27 }, { category: 'Category B', value: 25 }].")
            .nonempty({ message: "Pie chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        innerRadius: zod_1.z
            .number()
            .default(0)
            .describe("Set the innerRadius of pie chart, the value between 0 and 1. Set the pie chart as a donut chart. Set the value to 0.6 or number in [0 ,1] to enable it."),
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, innerRadius = 0, theme, title, width, outputType, } = params;
        // Transform data for ECharts
        const pieData = data.map((item) => ({
            name: item.category,
            value: item.value,
        }));
        const series = [
            {
                data: pieData,
                radius: innerRadius > 0 ? [`${innerRadius * 100}%`, "70%"] : "70%",
                type: "pie",
                emphasis: {
                    itemStyle: {
                        shadowBlur: 10,
                        shadowOffsetX: 0,
                        shadowColor: "rgba(0, 0, 0, 0.5)",
                    },
                },
            },
        ];
        const echartsOption = {
            legend: {
                left: "center",
                orient: "horizontal",
                top: "bottom",
            },
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "item",
                formatter: "{a} <br/>{b}: {c} ({d}%)",
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_pie_chart");
    }),
};
