// mcp-echarts/build/tools/sunburst.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSunburstChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Define recursive schema for hierarchical data
const SunburstNodeSchema = zod_1.z.lazy(() => zod_1.z.object({
    name: zod_1.z.string().describe("Node name, such as 'Technology'."),
    value: zod_1.z.number().describe("Node value, such as 100."),
    children: zod_1.z
        .array(SunburstNodeSchema)
        .optional()
        .describe("Child nodes for hierarchical structure."),
}));
exports.generateSunburstChartTool = {
    name: "generate_sunburst_chart",
    description: "Generate a sunburst chart to display multi-level hierarchical data, such as, organizational structure, file system hierarchy, or category breakdown.",
    inputSchema: zod_1.z.object({
        data: zod_1.z
            .array(SunburstNodeSchema)
            .describe("Data for sunburst chart, such as, [{ name: 'Technology', value: 100, children: [{ name: 'Frontend', value: 60, children: [{ name: 'React', value: 30 }] }] }].")
            .nonempty({ message: "Sunburst chart data cannot be empty." }),
        height: schema_1.HeightSchema,
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, theme, title, width, outputType } = params;
        const series = [
            {
                type: "sunburst",
                data: data,
                radius: [0, "90%"],
                center: ["50%", "50%"],
                sort: undefined,
                emphasis: {
                    focus: "ancestor",
                },
                label: {
                    show: true,
                    fontSize: 12,
                    color: "#000",
                    minAngle: 10,
                },
                itemStyle: {
                    borderRadius: 7,
                    borderWidth: 2,
                    borderColor: "#fff",
                },
                levels: [
                    {},
                    {
                        r0: "15%",
                        r: "35%",
                        itemStyle: {
                            borderWidth: 2,
                        },
                        label: {
                            rotate: "tangential",
                        },
                    },
                    {
                        r0: "35%",
                        r: "70%",
                        label: {
                            align: "right",
                        },
                    },
                    {
                        r0: "70%",
                        r: "72%",
                        label: {
                            position: "outside",
                            padding: 3,
                            silent: false,
                        },
                        itemStyle: {
                            borderWidth: 3,
                        },
                    },
                ],
            },
        ];
        const echartsOption = {
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "item",
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_sunburst_chart");
    }),
};
