// mcp-echarts/build/tools/tree.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTreeChartTool = void 0;
const zod_1 = require("zod");
const utils_1 = require("../utils/index.js");
const schema_1 = require("../utils/schema.js");
// Define recursive schema for hierarchical data
const TreeNodeSchema = zod_1.z.lazy(() => zod_1.z.object({
    name: zod_1.z.string().describe("Node name, such as 'Root'."),
    value: zod_1.z.number().optional().describe("Node value (optional)."),
    children: zod_1.z
        .array(TreeNodeSchema)
        .optional()
        .describe("Child nodes for hierarchical structure."),
}));
exports.generateTreeChartTool = {
    name: "generate_tree_chart",
    description: "Generate a tree chart to display hierarchical data structure, such as, organizational chart, family tree, or file directory structure.",
    inputSchema: zod_1.z.object({
        data: TreeNodeSchema.describe("Tree data structure, such as, { name: 'Root', children: [{ name: 'Child 1' }, { name: 'Child 2' }] }."),
        height: schema_1.HeightSchema,
        layout: zod_1.z
            .enum(["orthogonal", "radial"])
            .optional()
            .default("orthogonal")
            .describe("Tree layout type. Default is 'orthogonal'."),
        orient: zod_1.z
            .enum(["LR", "RL", "TB", "BT"])
            .optional()
            .default("LR")
            .describe("Tree orientation. LR=left-to-right, RL=right-to-left, TB=top-to-bottom, BT=bottom-to-top. Default is 'LR'."),
        theme: schema_1.ThemeSchema,
        title: schema_1.TitleSchema,
        width: schema_1.WidthSchema,
        outputType: schema_1.OutputTypeSchema,
    }),
    run: (params) => __awaiter(void 0, void 0, void 0, function* () {
        const { data, height, layout = "orthogonal", orient = "LR", theme, title, width, outputType, } = params;
        const series = [
            {
                type: "tree",
                data: [data],
                layout: layout,
                orient: orient,
                symbol: "emptyCircle",
                symbolSize: 7,
                initialTreeDepth: -1,
                itemStyle: {
                    color: "#4154f3",
                    borderWidth: 2,
                },
                lineStyle: {
                    color: "#ccc",
                    width: 1.5,
                    curveness: 0.5,
                },
                label: {
                    position: layout === "radial"
                        ? "top"
                        : orient === "LR"
                            ? "right"
                            : orient === "RL"
                                ? "left"
                                : orient === "TB"
                                    ? "bottom"
                                    : "top",
                    verticalAlign: "middle",
                    align: layout === "radial"
                        ? "center"
                        : orient === "LR"
                            ? "left"
                            : orient === "RL"
                                ? "right"
                                : "center",
                    fontSize: 12,
                },
                leaves: {
                    label: {
                        position: layout === "radial"
                            ? "top"
                            : orient === "LR"
                                ? "right"
                                : orient === "RL"
                                    ? "left"
                                    : orient === "TB"
                                        ? "bottom"
                                        : "top",
                        verticalAlign: "middle",
                        align: layout === "radial"
                            ? "center"
                            : orient === "LR"
                                ? "left"
                                : orient === "RL"
                                    ? "right"
                                    : "center",
                    },
                },
                emphasis: {
                    focus: "descendant",
                },
                expandAndCollapse: true,
                animationDuration: 550,
                animationDurationUpdate: 750,
            },
        ];
        const echartsOption = {
            series,
            title: {
                left: "center",
                text: title,
            },
            tooltip: {
                trigger: "item",
                triggerOn: "mousemove",
            },
        };
        return yield (0, utils_1.generateChartImage)(echartsOption, width, height, theme, outputType, "generate_tree_chart");
    }),
};
