// mcp-echarts/build/utils/imageHandler.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateChartImage = generateChartImage;
const minio_1 = require("./minio.js");
const render_1 = require("./render.js");
/**
 * Unified chart image generation method
 * Automatically decides whether to return Base64 image data or MinIO URL based on configuration
 *
 * @param echartsOption ECharts configuration options
 * @param width Image width, default 800
 * @param height Image height, default 600
 * @param theme Theme, default 'default'
 * @param outputType Output type, default 'png'
 * @param toolName Tool name (for debug logging)
 * @returns Unified MCP response content format
 */
function generateChartImage(echartsOption_1) {
    return __awaiter(this, arguments, void 0, function* (echartsOption, width = 800, height = 600, theme = "default", outputType = "png", toolName = "unknown") {
        // Debug logging
        if (process.env.DEBUG_MCP_ECHARTS) {
            console.error(`[DEBUG] ${toolName} generating chart:`, {
                width,
                height,
                theme,
                outputType,
                optionKeys: Object.keys(echartsOption),
            });
        }
        try {
            // Render chart
            const result = yield (0, render_1.renderECharts)(echartsOption, width, height, theme, outputType);
            // Determine output type
            const isImage = outputType !== "svg" && outputType !== "option";
            if (!isImage) {
                // SVG or configuration options, return text directly
                const response = {
                    content: [
                        {
                            type: "text",
                            text: result,
                        },
                    ],
                };
                if (process.env.DEBUG_MCP_ECHARTS) {
                    console.error(`[DEBUG] ${toolName} chart generated successfully:`, {
                        contentType: "text",
                        textLength: result.length,
                    });
                }
                return response;
            }
            // PNG image type
            const buffer = result;
            if ((0, minio_1.isMinIOConfigured)()) {
                try {
                    // Use MinIO storage, return URL
                    const url = yield (0, minio_1.storeBufferToMinIO)(buffer, "png", "image/png");
                    const response = {
                        content: [
                            {
                                type: "text",
                                text: url,
                            },
                        ],
                    };
                    if (process.env.DEBUG_MCP_ECHARTS) {
                        console.error(`[DEBUG] ${toolName} chart generated successfully:`, {
                            contentType: "text",
                            url: url,
                        });
                    }
                    return response;
                }
                catch (minioError) {
                    // MinIO failed, log warning and fallback to Base64
                    if (process.env.DEBUG_MCP_ECHARTS) {
                        console.error(`[DEBUG] ${toolName} MinIO storage failed, falling back to Base64:`, {
                            error: minioError instanceof Error
                                ? minioError.message
                                : String(minioError),
                        });
                    }
                    // Continue to Base64 fallback below
                }
            }
            // Fallback to Base64
            const base64Data = buffer.toString("base64");
            const response = {
                content: [
                    {
                        type: "image",
                        data: base64Data,
                        mimeType: "image/png",
                    },
                ],
            };
            if (process.env.DEBUG_MCP_ECHARTS) {
                console.error(`[DEBUG] ${toolName} chart generated successfully:`, {
                    contentType: "image",
                    dataLength: base64Data.length,
                });
            }
            return response;
        }
        catch (error) {
            // Error logging
            if (process.env.DEBUG_MCP_ECHARTS) {
                console.error(`[DEBUG] ${toolName} chart generation failed:`, {
                    error: error instanceof Error ? error.message : String(error),
                    stack: error instanceof Error ? error.stack : undefined,
                });
            }
            throw new Error(`Chart rendering failed: ${error instanceof Error ? error.message : String(error)}`);
        }
    });
}
