// mcp-echarts/build/utils/render.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderECharts = renderECharts;
const node_path_1 = __importDefault(require("node:path"));
const canvas_1 = require("@napi-rs/canvas");
const echarts = __importStar(require("echarts"));
const fontPath = node_path_1.default.join(__dirname, "..", "..", "fonts", "AlibabaPuHuiTi-3-55-Regular.otf");
canvas_1.GlobalFonts.registerFromPath(fontPath, "sans-serif");
/**
 * Render ECharts chart, return Buffer or string
 * This is a pure function that doesn't handle storage logic
 */
function renderECharts(echartsOption_1) {
    return __awaiter(this, arguments, void 0, function* (echartsOption, width = 800, height = 600, theme = "default", outputType = "png") {
        if (outputType === "svg" || outputType === "option") {
            const chart = echarts.init(null, theme, {
                renderer: "svg",
                ssr: true,
                width,
                height,
            });
            chart.setOption(Object.assign(Object.assign({}, echartsOption), { animation: false }));
            // Output string
            const svgStr = chart.renderToSVGString();
            // If the chart is no longer needed, call dispose to free memory
            chart.dispose();
            // Return SVG string or validated ECharts configuration options
            return outputType === "svg"
                ? svgStr
                : JSON.stringify(echartsOption, null, 2);
        }
        // Other output types (such as PNG) need to use Canvas
        const canvas = (0, canvas_1.createCanvas)(width, height);
        const chart = echarts.init(canvas, theme, {
            devicePixelRatio: 3,
        });
        echarts.setPlatformAPI({
            loadImage(src, onload, onerror) {
                const img = new Image();
                img.onload = onload.bind(img);
                img.onerror = onerror.bind(img);
                img.src = src;
                return img;
            },
        });
        chart.setOption(Object.assign(Object.assign({}, echartsOption), { animation: false }));
        // @ts-ignore
        const buffer = canvas.toBuffer("image/png");
        chart.dispose();
        return buffer;
    });
}
