// mcp-echarts/build/utils/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zodToJsonSchema = exports.WidthSchema = exports.TitleSchema = exports.OutputTypeSchema = exports.ThemeSchema = exports.HeightSchema = exports.AxisYTitleSchema = exports.AxisXTitleSchema = void 0;
const zod_1 = require("zod");
const zod_to_json_schema_1 = require("zod-to-json-schema");
exports.AxisXTitleSchema = zod_1.z
    .string()
    .optional()
    .default("")
    .describe("Set the x-axis title of chart.");
exports.AxisYTitleSchema = zod_1.z
    .string()
    .optional()
    .default("")
    .describe("Set the y-axis title of chart.");
exports.HeightSchema = zod_1.z
    .number()
    .int()
    .positive()
    .optional()
    .default(600)
    .describe("Set the height of the chart, default is 600px.");
exports.ThemeSchema = zod_1.z
    .enum(["default", "dark"])
    .optional()
    .default("default")
    .describe("Set the theme for the chart, optional, default is 'default'.");
exports.OutputTypeSchema = zod_1.z
    .enum(["png", "svg", "option"])
    .optional()
    .default("png")
    .describe("The output type of the diagram. Can be 'png', 'svg' or 'option'. Default is 'png', 'png' will return the rendered PNG image, 'svg' will return the rendered SVG string, and 'option' will return the valid ECharts option.");
exports.TitleSchema = zod_1.z
    .string()
    .optional()
    .describe("Set the title of the chart.");
exports.WidthSchema = zod_1.z
    .number()
    .int()
    .positive()
    .optional()
    .default(800)
    .describe("Set the width of the chart, default is 800px.");
// TODO: use zod v4 JSON to schema to replace zod-to-json-schema when v4 is stable
// biome-ignore lint/suspicious/noExplicitAny: <explanation>
const zodToJsonSchema = (schema) => {
    return (0, zod_to_json_schema_1.zodToJsonSchema)(schema, {
        rejectedAdditionalProperties: undefined,
    });
};
exports.zodToJsonSchema = zodToJsonSchema;
